agent/consul/state/config_entry.go

```diff
@@ -371,6 +373,100 @@ var serviceGraphKinds = []string{
 	structs.ServiceResolver,
 }
 
+// discoveryChainTargets will return a list of services listed as a target for the input's discovery chain
+func (s *Store) discoveryChainTargets(ws memdb.WatchSet, dc, service string, entMeta *structs.EnterpriseMeta) (uint64, []structs.ServiceName, error) {
+	source := structs.NewServiceName(service, entMeta)
+	req := discoverychain.CompileRequest{
+		ServiceName:         source.Name,
+		EvaluateInNamespace: source.NamespaceOrDefault(),
+
+		// TODO(freddy) : Should these be anything other than the known DC?
+		EvaluateInDatacenter: dc,
+		UseInDatacenter:      dc,
+	}
+	idx, chain, err := s.ServiceDiscoveryChain(ws, source.Name, entMeta, req)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to fetch discovery chain for %q: %v", source.String(), err)
+	}
+
+	var resp []structs.ServiceName
+	for _, t := range chain.Targets {
+		em := structs.EnterpriseMetaInitializer(t.Namespace)
+		target := structs.NewServiceName(t.Service, &em)
+
+		// TODO (freddy): Allow upstream DC and encode in response
+		if t.Datacenter == dc {
+			resp = append(resp, target)
+		}
+	}
+	return idx, resp, nil
+}
+
+// discoveryChainSources will return a list of services whose discovery chains have the input service as a target
+func (s *Store) discoveryChainSources(ws memdb.WatchSet, tx ReadTxn, dc string, destination structs.ServiceName) (uint64, []structs.ServiceName, error) {
+	queue := []structs.ServiceName{destination}
+
+	seenLink := make(map[structs.ServiceName]bool)
+	for len(queue) > 0 {
+		// The "link" index returns config entries that reference a service
+		iter, err := tx.Get(configTableName, "link", queue[0].ToServiceID())
+		if err != nil {
+			return 0, nil, err
+		}
+		ws.Add(iter.WatchCh())
+
+		for raw := iter.Next(); raw != nil; raw = iter.Next() {
+			entry := raw.(structs.ConfigEntry)
+
+			sn := structs.NewServiceName(entry.GetName(), entry.GetEnterpriseMeta())
+			if !seenLink[sn] {
+				seenLink[sn] = true
+				queue = append(queue, sn)
+			}
+		}
+		queue = queue[1:]
+	}
+
+	var (
+		maxIdx uint64
+		resp   []structs.ServiceName
+	)
+
+	// Only return the services that directly target the destination
```

Comment:
This comment misled me. I think what you meant is
```suggestion
	// Only return the services that target the destination anywhere in their discovery chains.
```

Reply:
The `"link"` index dump to me is a "direct" reference as your initial comment read.

Reply:
Fixed